### PR TITLE
Add with_mpi option to code setup.

### LINKED ISSUE #3236: Verdi code setup: add `withmpi` or `mpi_support` option
When setting up a code on a given computer it is already known to the user whether the code has been compiled with MPI support or not.

However, an AiiDA lab app that was build on top of AiiDA does not know this. Therefore, it does not know whether the option `withmpi` should be set to `True` or to `False` and can accidentally specify something like `mpirun -n 2 code.x` while the code was not compiled with an mpi support.

Of course this could be bypassed by using `mpirun -n 1 code.x`, which works also with the codes that are not compiled with mpi. But such a solution looks sub-optimal to me. Imagine an automated tool that automatically determines the number of cpus needed. How could it know that the code was not compiled with mpi support?

### PR DESCRIPTION
**Pull request by yakutovicha**
fixes #3236 

This PR allows specifying whether a code supports MPI.

### CODE REVIEW

**Review by sphuber — COMMENTED** [review]
Although it is good to have the possibility to define this on a `Code` instance, should there also not be some kind of enforcement in AiiDA? I understand that you mostly added this from the perspective of AiiDAlab where an app needs to introspect a Code to see what resources to specify, but shouldn't we then also not add validation in the `CalcJob` to prevent conflicting inputs from being specified? For example, if a `CalcJob` receives a `Code` with `withmpi = False` but the `metadata.options.withmpi` is set to `True`, what should happen? Should the option be ignored or should the input validation raise?

**Comment by yakutovicha** [comment]
> Although it is good to have the possibility to define this on a `Code` instance, should there also not be some kind of enforcement in AiiDA? I understand that you mostly added this from the perspective of AiiDAlab where an app needs to introspect a Code to see what resources to specify...

It came up in the AiiDAlab context, that is true. But I think the application of this goes slightly wider. Currently, one cannot define that certain codes need to always use MPI, while others don't. If this PR is merged, we would be able to do so.

> ... but shouldn't we then also not add validation in the `CalcJob` to prevent conflicting inputs from being specified? For example, if a `CalcJob` receives a `Code` with `withmpi = False` but the `metadata.options.withmpi` is set to `True`, what should happen? Should the option be ignored or should the input validation raise?

I think the way it is implemented currently makes sense: `metadata.options.withmpi` takes precedence. If for any weird reason the user wants to changes the default setting for a code, they can perfectly do that. This approach also avoids problems with backwards incompatibility, IMO.

**Comment by sphuber** [comment]
> I think the way it is implemented currently makes sense: `metadata.options.withmpi` takes precedence.

That suggests that there _is_ any interaction between the two and the metadata option is leading. But I don't see where in `aiida-core` the `Code.get_withmpi` is used at all. Am I missing something?

**Comment by yakutovicha** [comment]
> That suggests that there _is_ any interaction between the two and the metadata option is leading. But I don't see where in `aiida-core` the `Code.get_withmpi` is used at all. Am I missing something?

You are right. I've just addressed that.

**Comment by unkcpz** [comment]
My first question is whether this should be a flag? It is obvious that when a code only run in non-mpi mode the option should be set to `False`. But when this option set to `withmpi=True` is that always means the code can be run at mpi mode and serial mode. Is there any situation that code only run in mpi mode and will fail in serial mode?

> I think the way it is implemented currently makes sense: metadata.options.withmpi takes precedence. If for any weird reason the user wants to changes the default setting for a code, they can perfectly do that. This approach also avoids problems with backwards incompatibility, IMO.

To make it easy to understand, I think we have three places where the `withmpi` is setting, classify the users into following three: 
- [code setup by user]: as @yakutovicha try to address in this PR
- [plugin developer]: `code_info.withmpi` which normally set by the plugin developer to indicate how the code should be run.
- [calcjob running user]: `calcjob.metadata.withmpi` which set by the end user of `CalcJob` set how they going to run the calcjob process.

Usually, the code setup user and calcjob running user are same person, and we should assume that plugin developer know nothing about how the code will actually compiled at the computer. When plugin developer set the option they must have strong opinion about the code only run in specific mode otherwise they just leave `code_info.withmpi` unset and it will be set through `calcjob.metadata.withmpi` by calcjob running user. 

However, I kindly have opposite opinion that we should disallowed users to override the mpi setting of code by `code_info.withmpi` and `metadata.options.withmpi`. I think the `verdi code setup` stage is more fixed since the information is store in the database once for all, therefore users need and will take more consideration on setting this and always know how the code was compiled and how the code should be used. 
So when conflict between code setup and plugin `code_info.withmpi` it is always show a potential issue(setting imcompatible) between them. The user should contact plugin developer to deal with the problem.

**Review by unkcpz — COMMENTED** [review]
My current point is this is a great option to have for code setup but we need more discussion in the priorities of how different `withmpi` setting take effect at final place.

**Comment by yakutovicha** [comment]
> My first question is whether this should be a flag?

I guess it should be a flag. Or, in other words, I don't understand why it shouldn't.

> It is obvious that when a code only run in non-mpi mode the option should be set to `False`. But when this option set to `withmpi=True` is that always means the code can be run at mpi mode and serial mode.

That is definitely true.

> Is there any situation that code only run in mpi mode and will fail in serial mode?

I don't think so. What I am trying to handle in this PR is the situations when a parallel code is compiled in a serial mode. In such a situation the user can configure a code and can explicitly say that `my-serial-code@localhost` should not run with MPI.



> To make it easy to understand, I think we have three places where the `withmpi` is setting, classify the users into following three:
> 
> * [code setup by user]: as @yakutovicha try to address in this PR
> * [plugin developer]: `code_info.withmpi` which normally set by the plugin developer to indicate how the code should be run.
> * [calcjob running user]: `calcjob.metadata.withmpi` which set by the end user of `CalcJob` set how they going to run the calcjob process.


Fully correct. The only thing I would like to add here is the **current** priority (from lower to higher):
1. Code setup by the user.
2. Calcjob running user.
3. Plugin developer.

> > I think the way it is implemented currently makes sense: metadata.options.withmpi takes precedence. If for any weird reason the user wants to changes the default setting for a code, they can perfectly do that. This approach also avoids problems with backwards incompatibility, IMO.
> 
> However, I kindly have opposite opinion that we should disallowed users to override the mpi setting of code by `code_info.withmpi` and `metadata.options.withmpi`

When I wrote the above, I was referring to "code setup by the user" and "calcjob running the user". I think the latter should be prioritised and this is what I wrote. I didn't think about the "plugin developer", but I trust your choice.

> I think the `verdi code setup` stage is more fixed since the information is store in the database once for all, therefore users need and will take more consideration on setting this and always know how the code was compiled and how the code should be used.

They should definitely know that. To me, knowing whether the code is serial or parallel is one of the most basic things about a code.

**Review by unkcpz — APPROVED** [review]
Thanks for explanation @yakutovicha, I should take a look at the changed code before I raise so many questions, the answers already at the code. 

All changes looks good to me, but have no idea why the tests are failed.

**Comment by unkcpz** [comment]
Hi @yakutovicha , it seems other tests failed after you fix the tests. Could you rebase to the develop and run the action again? Thanks!

File: aiida/cmdline/params/options/commands/code.py
Comment (sphuber) [inline aiida/cmdline/params/options/commands/code.py:95]: For clarity I think it is always better to separate words with hyphens
```suggestion
    '--with-mpi/--without-mpi',
```

File: aiida/orm/nodes/data/code.py
Comment (sphuber) [inline aiida/orm/nodes/data/code.py:331]: Should really add a type check here and raise `TypeError` if incorrect. Something like:
```python
from aiida.common.lang import type_check
type_check(withmpi, bool)
```

**Review by sphuber — CHANGES_REQUESTED** [review]
I realize I am late to the party, but I just read through your conversation on determining the priority and I am not sure that I agree. I have put a comment in the code explaining why and what I would do instead. If you don't agree, maybe we schedule a quick meeting to hash it out.

File: aiida/engine/processes/calcjobs/calcjob.py
Comment (yakutovicha) [reply aiida/engine/processes/calcjobs/calcjob.py:707]: I agree with @sphuber's suggestion. If @unkcpz is not against it - I am happy to modify the logic according to it.

Comment (unkcpz) [reply aiida/engine/processes/calcjobs/calcjob.py:707]: Fair enough, I have no objection to this. 

>  If this really is the case, then instead of ignoring the other settings, I think the plugin should simply add a check in the prepare_for_submission and raise an error.

Only for this, is this mean the code for checking needs to be in the plugin, and the plugin developer need to do it explicitly?

Comment (yakutovicha) [reply aiida/engine/processes/calcjobs/calcjob.py:707]: > >  If this really is the case, then instead of ignoring the other settings, I think the plugin should simply add a check in the prepare_for_submission and raise an error.

> Only for this, is this mean the code for checking needs to be in the plugin, and the plugin developer need to do it explicitly?

I think this is what @sphuber meant. And it makes sense to me. The case is so special, so it probably needs special treatment.

Comment (sphuber) [reply aiida/engine/processes/calcjobs/calcjob.py:707]: Yes that is indeed what I meant. If a plugin should really never ever be run in parallel, then the plugin developer should simply put in a check in `prepare_for_submission` and raise if either `Code.withmpi` or `metadata.options.withmpi` is `True`.

Comment (yakutovicha) [reply aiida/engine/processes/calcjobs/calcjob.py:707]: Ok, since we agreed, I am going to proceed to implement the discussed changes.

Comment (yakutovicha) [reply aiida/engine/processes/calcjobs/calcjob.py:707]: I just realised that since we are going to set `withmpi` option in the code setup, and since it has higher priority than the plugin - there is no need to keep this option in the plugin.

Comment (unkcpz) [reply aiida/engine/processes/calcjobs/calcjob.py:707]: the code setup is made by the user of the plugin, we need a way to let the plugin developers to set the default mpi behavior of the code. But you are right, if the `withmpi` is a must set parameter for the code in code setup, it will always override the plugin set and leave it not being used. The is a problem.

Comment (sphuber) [reply aiida/engine/processes/calcjobs/calcjob.py:707]: That is what I described in my last comment though: if there really is a conflict and a plugin can never be run with a specific value of `withmpi` and either the `Code` value or `metadata.option.withmpi` specifies the wrong value. The plugin can simply raise a validation error. The user will then simply have to change the input, or change `Code` that has the correct withmpi setting.

File: aiida/orm/nodes/data/code.py
Comment (yakutovicha) [reply aiida/orm/nodes/data/code.py]: done.

**Comment by yakutovicha** [comment]
I finished the implementation. @sphuber and @unkcpz can you please have a look before I fix the tests?

Another thing: I would prefer to rename `withmpi` option to `with_mpi` are there any objections to it? I think `with_mpi` is more in line with the other options.

File: aiida/cmdline/params/options/commands/code.py
Comment (unkcpz) [inline aiida/cmdline/params/options/commands/code.py:143]:
```suggestion
    default=True,
    required_fn=is_on_computer,
    prompt_fn=is_on_computer,
```
I think it is more make sense to only set this for the remote code. Compiling with mpi usually link to the library therefore for the local code we should always run in serial. Also check https://www.open-mpi.org/faq/?category=mpi-apps#static-mpi-apps . It says 'Can I build 100% static MPI applications? Fully static linking is not for the weak, and it is not recommended. But it is possible, with some caveats.'

**Review by unkcpz — CHANGES_REQUESTED** [review]
@yakutovicha Thanks! Just one minor request. All looks good to me.

File: aiida/cmdline/params/options/commands/code.py
Comment (yakutovicha) [reply aiida/cmdline/params/options/commands/code.py:143]: It is actually possible to run python scripts with mpirun: https://rabernat.github.io/research_computing/parallel-programming-with-mpi-for-python.html

**Comment by ltalirz** [comment]
Hey guys, I think this addition is very useful, so I wanted to check about the status of this PR and add my thoughts.

To recap: `with_mpi=True` can occur at three different levels, with slightly different meanings:

  - Calculation level: User says "run this calculation with MPI"
  - CalcJob plugin level: Plugin developer says "run these types of CalcJobs with MPI by default"
  - Code level: User / HPC admin says "this code should be run with MPI"

Given this, here is how I would expect these levels to interact as a user:
  - The setting at the Code level should default to something like `None` (="unknown"). This is also the only way we can migrate old codes in a backward-compatible way.
  - The setting at the Code level (if specified) will take precedence over the default value specified at the plugin level
  - If a setting at the Code level and a setting at the calculation level conflict, AiiDA should except (the user is trying to do something that is not supported).

Do you agre?


Finally, as much as I prefer `with_mpi` over `withmpi`, this is not worth breaking a large number of AiiDA plugins and user scripts. Feel free to choose the name at the code level, but I strongly advise against changing the existing names at the plugin and calculation levels.

**Comment by ltalirz** [comment]
P.S. I just realized that level B) is further split into two cases:
The one I had in mind, say B.1, was specifying a default value in the `CalJob` spec
```python
     spec.input("metadata.options.withmpi", valid_type=bool, default=True)
```
As e.g. done [here](https://github.com/aiidateam/aiida-cp2k/blob/main/aiida_cp2k/calculations/__init__.py#L131)

However, strictly speaking, there is another case, let's call it B.2: hardcoding `codeinfo.withmpi = False` in the `prepare_for_submission` function, e.g. [here](https://github.com/ltalirz/aiida-zeopp/blob/master/aiida_zeopp/calculations/network.py#L104).
This is the plugin developer saying "I know this code always/never supports MPI".
In terms of  precedence, it should therefore be treated like the code level (C). If there is a conflict with the code-level specification, AiiDA should except.

**Comment by yakutovicha** [comment]
@ltalirz I didn't manage to merge this PR when I was at the AiiDA retreat and in the upcoming month I am not planning to work on it (busy with the AiiDAlab migration). Feel free to take it over, if you have time

**Comment by sphuber** [comment]
I will try and have a look soon to revive and finalize this

**Comment by yakutovicha** [comment]
> I will try and have a look soon to revive and finalize this

thanks 👍

**Comment by sphuber** [comment]
Superseded by #5922

**Comment by sphuber** [comment]
Superseded by #5922

**Pull request closed without merging**
